File: app/src/main/java/com/nacare/capture/ui/login/LoginActivity.java
package com.nacare.capture.ui.login;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.Html;
import android.text.SpannableString;
import android.text.Spanned;
import android.text.TextWatcher;
import android.text.method.LinkMovementMethod;
import android.text.style.ForegroundColorSpan;
import android.text.style.UnderlineSpan;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;

import com.nacare.capture.R;
import com.nacare.capture.data.Sdk;
import com.nacare.capture.data.model.FormatterClass;
import com.nacare.capture.data.service.ActivityStarter;
import com.nacare.capture.ui.main.MainActivity;
import com.nacare.capture.ui.main.SyncActivity;
import com.nacare.capture.ui.programs.ProgramsActivity;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;

import io.reactivex.disposables.Disposable;

public class LoginActivity extends AppCompatActivity {

    private LoginViewModel loginViewModel;
    private Disposable disposable;

    private TextInputEditText serverUrlEditText;
    private TextInputEditText usernameEditText;
    private TextInputEditText passwordEditText;
    private MaterialButton loginButton;
    private TextView textView, mailer;


    public static Intent getLoginActivityIntent(Context context) {
        return new Intent(context, LoginActivity.class);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        loginViewModel = new ViewModelProvider(this, new LoginViewModelFactory()).get(LoginViewModel.class);

        serverUrlEditText = findViewById(R.id.serverUrlEdittext);
        usernameEditText = findViewById(R.id.usernameEdittext);
        passwordEditText = findViewById(R.id.passwordEdittext);
        loginButton = findViewById(R.id.loginButton);

        String recover = "<a href=\"https://nacareke.on.spiceworks.com/portal/registrations\"><u>Account Recovery</u></a>";
        String mail = "For assistance on the National Cancer <br>Registry of Kenya System, click here or send an email to<br><br> <a href=\"mailto:help@nacare.on.spiceworks.com\">help@nacare.on.spiceworks.com</a>";
        textView = findViewById(R.id.tv_recovery);
        mailer = findViewById(R.id.tv_mailer);

        // Use Html.fromHtml() to interpret the HTML formatting
        SpannableString span = new SpannableString(Html.fromHtml(recover));

        int start_span = span.toString().indexOf("Account Recovery");
        int end_span = start_span + "Account Recovery".length();

        span.setSpan(new ForegroundColorSpan(getResources().getColor(android.R.color.white)), start_span, end_span, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
        span.setSpan(new UnderlineSpan(), start_span, end_span, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
        textView.setText(span);
        textView.setMovementMethod(LinkMovementMethod.getInstance());

        // Use Html.fromHtml() to interpret the HTML formatting
        SpannableString spannableString = new SpannableString(Html.fromHtml(mail));

        // Customize the appearance of the link (white color and underline)
        int start = spannableString.toString().indexOf("help@nacare.on.spiceworks.com");
        int end = start + "help@nacare.on.spiceworks.com".length();

        spannableString.setSpan(new ForegroundColorSpan(getResources().getColor(android.R.color.white)), start, end, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);
        spannableString.setSpan(new UnderlineSpan(), start, end, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);

        mailer.setText(spannableString);
        mailer.setMovementMethod(LinkMovementMethod.getInstance());

        loginViewModel.getLoginFormState().observe(this, loginFormState -> {
            if (loginFormState == null) {
                return;
            }
            loginButton.setEnabled(loginFormState.isDataValid());
            if (loginFormState.getServerUrlError() != null) {
                serverUrlEditText.setError(getString(loginFormState.getServerUrlError()));
            }
            if (loginFormState.getUsernameError() != null) {
                usernameEditText.setError(getString(loginFormState.getUsernameError()));
            }
            if (loginFormState.getPasswordError() != null) {
                passwordEditText.setError(getString(loginFormState.getPasswordError()));
            }
        });

        loginViewModel.getLoginResult().observe(this, loginResult -> {
            if (loginResult == null) {
                return;
            }
            if (loginResult.getError() != null) {
                showLoginFailed(loginResult.getError());
            }
            if (loginResult.getSuccess() != null) {
                new FormatterClass().saveSharedPref("serverUrl", serverUrlEditText.getText().toString(), LoginActivity.this);
                new FormatterClass().saveSharedPref("username", usernameEditText.getText().toString(), LoginActivity.this);
                new FormatterClass().saveSharedPref("password", passwordEditText.getText().toString(), LoginActivity.this);
                ActivityStarter.startActivity(this, SyncActivity.getIntent(this), true);

            }
            setResult(Activity.RESULT_OK);
        });

        TextWatcher afterTextChangedListener = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                // ignore
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                // ignore
            }

            @Override
            public void afterTextChanged(Editable s) {
                loginViewModel.loginDataChanged(
                        serverUrlEditText.getText().toString(),
                        usernameEditText.getText().toString(),
                        passwordEditText.getText().toString());
            }
        };
        serverUrlEditText.setText("http://45.79.116.38:8080/");
        serverUrlEditText.addTextChangedListener(afterTextChangedListener);
        usernameEditText.addTextChangedListener(afterTextChangedListener);
        passwordEditText.addTextChangedListener(afterTextChangedListener);

        passwordEditText.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                login();
            }
            return false;
        });

        loginButton.setOnClickListener(v -> login());
    }

    private void login() {
        loginButton.setVisibility(View.INVISIBLE);
        String username = usernameEditText.getText().toString();
        String password = passwordEditText.getText().toString();
        String serverUrl = serverUrlEditText.getText().toString();

        disposable = loginViewModel
                .login(username, password, serverUrl)
                .doOnTerminate(() -> loginButton.setVisibility(View.VISIBLE))
                .subscribe(u -> {
                }, t -> {
                });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (disposable != null) {
            disposable.dispose();
        }
    }

    private void showLoginFailed(String errorString) {
        Toast.makeText(getApplicationContext(), errorString, Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/com/nacare/capture/auth/LoginActivity.kt
package com.nacare.capture.auth

import android.app.ProgressDialog
import android.content.Intent
import android.os.Bundle
import android.text.Html
import android.text.SpannableString
import android.text.Spanned
import android.text.method.LinkMovementMethod
import android.text.style.ForegroundColorSpan
import android.text.style.UnderlineSpan
import android.util.Base64
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.nacare.capture.MainActivity
import com.nacare.capture.R
import com.nacare.capture.data.FormatterClass
import com.nacare.capture.databinding.ActivityLoginBinding
import com.nacare.capture.network.RetrofitCalls

class LoginActivity : AppCompatActivity() {
    private val retrofitCalls = RetrofitCalls()
    private val formatter = FormatterClass()
    private lateinit var binding: ActivityLoginBinding
    private lateinit var progressDialog: ProgressDialog
    override fun onStart() {
        super.onStart()
        val login = formatter.getSharedPref("isLoggedIn", this@LoginActivity)
        if (login != null) {
            if (login == "true") {
                val intent = Intent(this@LoginActivity, MainActivity::class.java)
                startActivity(intent)
                finish()
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityLoginBinding.inflate(layoutInflater)
        setContentView(binding.root)

        progressDialog = ProgressDialog(this)
        progressDialog.setMessage("Please wait...") // Set your message
        progressDialog.setCancelable(true)
        binding.apply {

            val recover =
                "<a href=\"https://nacareke.on.spiceworks.com/portal/registrations\"><u>Account Recovery</u></a>"
            val mail =
                "For assistance on the National Cancer <br>Registry of Kenya System, click here or send an email to<br><br> <a href=\"mailto:help@nacare.on.spiceworks.com\">help@nacare.on.spiceworks.com</a>"
            val span = SpannableString(Html.fromHtml(recover))

            val start_span = span.toString().indexOf("Account Recovery")
            val end_span = start_span + "Account Recovery".length

            span.setSpan(
                ForegroundColorSpan(resources.getColor(R.color.white)),
                start_span,
                end_span,
                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            span.setSpan(UnderlineSpan(), start_span, end_span, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)

            tvRecovery.apply {
                text = span
                movementMethod = LinkMovementMethod.getInstance()
            }

            // Use Html.fromHtml() to interpret the HTML formatting
            // Use Html.fromHtml() to interpret the HTML formatting
            val spannableString = SpannableString(Html.fromHtml(mail))

            // Customize the appearance of the link (white color and underline)

            // Customize the appearance of the link (white color and underline)
            val start = spannableString.toString().indexOf("help@nacare.on.spiceworks.com")
            val end = start + "help@nacare.on.spiceworks.com".length

            spannableString.setSpan(
                ForegroundColorSpan(resources.getColor(R.color.white)),
                start,
                end,
                Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
            )
            spannableString.setSpan(UnderlineSpan(), start, end, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            tvMailer.apply {
                text = spannableString
                movementMethod = LinkMovementMethod.getInstance()
            }
            loginButton.apply {
                setOnClickListener {
                    validateData()
                }
            }
        }

//        retrofitCalls.signIn(this, "admin", "district")


    }

    private fun validateData() {
        val username = binding.usernameEdittext.text.toString()
        val password = binding.passwordEdittext.text.toString()
        if (username.isEmpty()) {
            Toast.makeText(this, "Enter Username", Toast.LENGTH_SHORT).show()
            binding.usernameEdittext.requestFocus()
            return
        }
        if (password.isEmpty()) {
            Toast.makeText(this, "Enter Password", Toast.LENGTH_SHORT).show()
            binding.passwordEdittext.requestFocus()
            return
        }

        val credentials = "${username.trim()}:${password.trim()}"

        val encodedString = Base64.encodeToString(credentials.toByteArray(), Base64.NO_WRAP)
        formatter.saveSharedPref("encodedString", encodedString, this@LoginActivity)
        retrofitCalls.signIn(this@LoginActivity, progressDialog)


    }
}